docs/input_output.md

```diff
@@ -2,85 +2,279 @@
+                                 // the workflow to determine next steps
+      "payload": "<object>"      // A Task-specific payload. This can be any data type required by the Task.
+                                 // It can be considered the input and output of the Task, whereas the other 
+                                 // fields are execution context. Tasks should document their expected payload 
+                                 // input and output formats. Generally a Task will return an object which is 
+                                 // nearly identical to its input in all fields but 'payload', and 'payload' 
+                                 // will be completely different"
+    }
+
+## Cumulus Message Adapter
 
-The Scheduler service creates the initial event by combining the collection configuration (See [collections.json](../config/collections.json)) with the AWS execution context provided by its [CloudFormation template](../config/cloudformation.yml.erb).  The collection configuration supports inheritance. If the `"parent"` attribute of a collection references the `"id"` of another collection, the other collection's fields will be recursively copied into the collection.
+The Scheduler service creates the initial event by combining the collection configuration, external resource configuration, workflow configuration, and deployment environment settings.  The subsequent workflow messages between tasks must conform to the message schema. By using Cumulus Message Adapter, individule task Lambda function can focus on the input and output specific to the task, and does not worry about the non-task related message fields.
 
-Because of the potential size of the `"payload"` field, it may contain a reference to an S3 Bucket and Key, as follows:
+Cumulus Message Adapter implements an AWS Lambda handler that adapts incoming messages in the Cumulus protocol to a format more easily consumable by Cumulus tasks, invokes the tasks, and then adapts their response back to the Cumulus message protocol to be sent to the next task.  
+
+A task Lambda function can be configured to use Cumulus Message Adapter to help construct the input/output messages and resolve task configurations. In the Lambda function configuration file, a task Lambda function can be configured to use Cumulus Message Adapter, for example,
+
+    DiscoverPdrs:
+      handler: index.handler
+      useMessageAdapter: true  
+
+A task Lambda function using the Message Adapter shall take a event message which contains two fields, `"input"` and `"config"`.
+> `input`: By default, the incoming payload is the payload from previous task, or it can be a portion of the payload as specified in the task configuration.
+
+> `config`: Task-specific configuration object with URL templates resolved.
+
+> `Task output`: By default, the task's return value is the next payload, or a portion of it is interpreted as the next payload as specified in the task configuration.
```

Comment:
I wouldn't include `Task output` as another point here to make it clear `input` and `config` are the input to the task's application code. I would just have a sentence which follows. Suggested format:

```
Input to task application code is a json object with keys:
* `"input":` By default, the incoming payload is the payload from previous task, or it can be a portion of the payload as specified in the task configuration.
* `"config":` Task-specific configuration object with URL templates resolved.

Output of the task application code:
* By default, the task's return value is the next `payload`, or a portion of it is interpreted as the next payload as specified in the task configuration.
```